File: android/src/main/kotlin/io/sentient/android/settings/advanced/AdvancedViewModel.kt
// ---------------------------------------------------------------------------
// AdvancedViewModel — Advanced settings page. SLOW save: a full-profile PUT whose
// diff touches compression/advanced (not audio), so ApplyProfileChangeUseCase PUTs
// then applies (Hermes restart). Holds `original` (server truth) + `draft`; dirty is
// derived. Thin: loads via profileRepository, saves via the one FSM usecase, folds
// its ApplyState flow into flat progress fields (Restarting drives the restart copy).
// ---------------------------------------------------------------------------
package io.sentient.android.settings.advanced

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.sentient.mobiledata.di.SettingsComponent
import io.sentient.mobiledata.result.SentientResult
import io.sentient.mobiledata.usecase.settings.ApplyState
import io.sentient.mobiledata.usecase.settings.ProfileMutation
import io.sentient.mobilesdk.log.createLogger
import io.sentient.mobilesdk.settings.ProfileV1
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/** Advanced page state. `original` is the last server truth; `draft` carries local edits. */
data class AdvancedUiState(
    val loading: Boolean = true,
    val loadError: String? = null,
    val original: ProfileV1? = null,
    val draft: ProfileV1? = null,
    val saving: Boolean = false,
    val restarting: Boolean = false,
    val alreadyApplying: Boolean = false,
    val applyError: String? = null,
) {
    val dirty: Boolean get() = original != null && draft != null && draft != original
    val applyActive: Boolean get() = saving || restarting
}

class AdvancedViewModel(private val component: SettingsComponent) : ViewModel() {
    private val log = createLogger("android", "settings", "advanced-vm")

    private val _ui = MutableStateFlow(AdvancedUiState())
    val ui: StateFlow<AdvancedUiState> = _ui.asStateFlow()

    init {
        load()
    }

    private fun load() {
        viewModelScope.launch {
            _ui.update { it.copy(loading = true, loadError = null) }
            when (val r = component.profileRepository.getProfile()) {
                is SentientResult.Success ->
                    _ui.update { it.copy(loading = false, original = r.data, draft = r.data) }
                is SentientResult.Failure -> {
                    log.warn("load.failed", mapOf("kind" to r.error.kind))
                    _ui.update { it.copy(loading = false, loadError = r.error.userMessage) }
                }
                is SentientResult.Loading -> Unit
            }
        }
    }

    fun setReasoning(effort: String) = editDraft { it.copy(advanced = it.advanced.copy(reasoningEffort = effort)) }

    fun setMaxTokens(tokens: Int) = editDraft { it.copy(advanced = it.advanced.copy(maxTokens = tokens)) }

    fun setExtraPrompt(text: String) = editDraft { it.copy(advanced = it.advanced.copy(extraSystemPrompt = text)) }

    fun setCompression(threshold: Double) =
        editDraft { it.copy(compression = it.compression.copy(threshold = threshold)) }

    private inline fun editDraft(transform: (ProfileV1) -> ProfileV1) {
        _ui.update { s ->
            val draft = s.draft ?: return@update s
            s.copy(draft = transform(draft), alreadyApplying = false, applyError = null)
        }
    }

    fun save() {
        val state = _ui.value
        val previous = state.original ?: return
        val next = state.draft ?: return
        if (!state.dirty || state.applyActive) return
        log.info("save", mapOf("reasoning" to next.advanced.reasoningEffort, "maxTokens" to next.advanced.maxTokens))
        viewModelScope.launch {
            component.applyProfileChange(ProfileMutation.PutProfile(previous, next)).collect { st ->
                _ui.update { it.foldApply(st) }
                if (st is ApplyState.Ready) refetch()
            }
        }
    }

    private suspend fun refetch() {
        when (val r = component.profileRepository.getProfile()) {
            is SentientResult.Success -> _ui.update { it.copy(original = r.data, draft = r.data) }
            is SentientResult.Failure -> log.warn("refetch.failed", mapOf("kind" to r.error.kind))
            is SentientResult.Loading -> Unit
        }
    }
}

/** Fold one FSM transition into flat progress fields. */
internal fun AdvancedUiState.foldApply(state: ApplyState): AdvancedUiState = when (state) {
    ApplyState.Saving -> copy(saving = true, restarting = false, alreadyApplying = false, applyError = null)
    ApplyState.Restarting -> copy(saving = false, restarting = true)
    is ApplyState.Ready -> copy(saving = false, restarting = false, alreadyApplying = false, applyError = null)
    ApplyState.AlreadyApplying -> copy(saving = false, restarting = false, alreadyApplying = true)
    is ApplyState.Failed -> copy(saving = false, restarting = false, applyError = state.error.userMessage)
    ApplyState.Idle -> this
}

File: ios/App/Settings/Advanced/AdvancedViewModel.swift
// ---------------------------------------------------------------------------
// AdvancedViewModel — the Advanced settings page state holder. SLOW save: the
// profile PUT changes non-audio fields, so ApplyProfileChangeUseCase runs the
// PUT-then-apply path that blocks through a Hermes worker restart.
//
// Holds the loaded ProfileV1 as `original` (the mutation's `previous`) plus the
// Swift-native draft primitives the controls bind to (reasoning / compression /
// max-tokens / extra-system-prompt). Dirty is derived from those primitives —
// KMP data classes aren't Swift-Equatable, so ProfileV1 is never compared here.
// ---------------------------------------------------------------------------
import Foundation
import MobileData

@MainActor
@Observable
final class AdvancedViewModel {
    enum Phase: Equatable {
        case loading
        case ready
        case failed(String)
    }

    enum Save: Equatable {
        case idle
        case saving
        case restarting
        case alreadyApplying
        case failed(String)
    }

    private(set) var phase: Phase = .loading
    private(set) var save: Save = .idle

    /// Draft primitives bound to the controls.
    var reasoningEffort = ProfileEnums.shared.reasoningEfforts.first ?? "minimal"
    var threshold: Double = 0
    var maxTokens: Double = 0
    var extraSystemPrompt = ""

    private var original: ProfileV1?
    private let settings: SettingsComponent
    private let log = AppLog("settings", "advanced-vm")

    init(settings: SettingsComponent) {
        self.settings = settings
    }

    var isDirty: Bool {
        guard let o = original else { return false }
        return reasoningEffort != o.advanced.reasoningEffort
            || threshold != o.compression.threshold
            || Int32(maxTokens.rounded()) != o.advanced.maxTokens
            || extraSystemPrompt != o.advanced.extraSystemPrompt
    }

    var isApplying: Bool { save == .saving || save == .restarting }

    func load() async {
        log.info("load")
        do {
            let result = try await settings.profileRepository.getProfile()
            switch onEnum(of: result) {
            case .success(let s):
                apply(s.data)
                phase = .ready
                log.info("load.ready reasoning=\(reasoningEffort) maxTokens=\(Int(maxTokens))")
            case .failure(let f):
                phase = .failed(f.error.userMessage)
                log.warn("load.failed kind=\(f.error.kind)")
            case .loading:
                phase = .loading
            }
        } catch is CancellationError {
        } catch {
            phase = .failed("Couldn't load advanced settings.")
            log.warn("load.threw")
        }
    }

    func save() async {
        guard let o = original, isDirty else { return }
        log.info("save.start reasoning=\(reasoningEffort)")
        let mutation = ProfileMutationPutProfile(previous: o, next: nextProfile(from: o))
        for await state in settings.applyProfileChange.invoke(mutation: mutation) {
            switch onEnum(of: state) {
            case .idle: break
            case .saving: save = .saving
            case .restarting: save = .restarting
            case .ready:
                save = .idle
                log.info("save.ready")
                await load()
            case .alreadyApplying:
                save = .alreadyApplying
                log.warn("save.already-applying")
            case .failed(let f):
                save = .failed(f.error.userMessage)
                log.warn("save.failed")
            }
        }
    }

    private func apply(_ profile: ProfileV1) {
        original = profile
        reasoningEffort = profile.advanced.reasoningEffort
        threshold = profile.compression.threshold
        maxTokens = Double(profile.advanced.maxTokens)
        extraSystemPrompt = profile.advanced.extraSystemPrompt
    }

    private func nextProfile(from o: ProfileV1) -> ProfileV1 {
        ProfileV1(
            schemaVersion: o.schemaVersion,
            userId: o.userId,
            model: o.model,
            voice: o.voice,
            audio: o.audio,
            persona: o.persona,
            tools: o.tools,
            compression: ProfileCompression(threshold: threshold),
            advanced: ProfileAdvanced(
                extraSystemPrompt: extraSystemPrompt,
                maxTokens: Int32(maxTokens.rounded()),
                reasoningEffort: reasoningEffort
            ),
            devices: o.devices
        )
    }
}
